--- a/Core/Internal/CollectionComponentOperations.cs
+++ b/Core/Internal/CollectionComponentOperations.cs
@@ -172,7 +172,7 @@
             if (needsAdd)
                 context->latiosWorld.m_impl->m_worldUnmanaged.EntityManager.AddComponent(context->addQuery, t);
             if (needsRemove)
-                context->latiosWorld.m_impl->m_worldUnmanaged.EntityManager.RemoveComponent(context->addQuery, t);
+                context->latiosWorld.m_impl->m_worldUnmanaged.EntityManager.RemoveComponent(context->removeQuery, t);
         }
 
         static void DisposeCollectionStorage<T>(DisposeCollectionStorageContext* context) where T : unmanaged, ICollectionComponent, StaticAPI.ICollectionComponentSourceGenerated
